### PR TITLE
[WIP] Fix of automatic thread pool adjustment in Twisted's ThreadedResolver

### PR DESCRIPTION
**Pull request by sibiryakov**
During massive broad crawls with high concurrency, sometimes DNS resolutions failing with timeouts. The reason is usage of Twisted ThreadedResolver, which has no automatic thread pool size adjustment capability. This patch introduces such capability among with adjustment of default Twisted thread pool size.

We try to adjust thread pool size each 20-th call, instead of every call. The motivation is IMO, there is risk that with further changes in Twisted this code could cause joining/spinning up of new thread on every call, which will affect performance.

In Twisted the code affected is here https://github.com/twisted/twisted/blob/trunk/twisted/python/threadpool.py#L239 This function called only during initialization, and never during operation.

### CODE REVIEW

**Comment by sibiryakov** [comment]
Hey @pablohoffman please take a look.

**Comment by sibiryakov** [comment]
@kmike ping.

File: scrapy/settings/default_settings.py [hunk 1]
```diff
@@ -52,6 +52,9 @@
 DEPTH_PRIORITY = 0
 
 DNSCACHE_ENABLED = True
+DNS_CACHE_SIZE = 10000
```
Comment (kmike) [inline scrapy/settings/default_settings.py]: DNSCACHE_SIZE is better because we already have DNSCACHE_ENABLED

**Comment by kmike** [comment]
The problem with adjustPoolSize / DNS_MAX_THREADS is that ThreadedResolver uses a global thread pool, so DNS_MAX_THREADS affects file pipelines, storages and all other code which uses deferToThread.

**Comment by sibiryakov** [comment]
Ok, it seems adjusting threadPool size affects too many parts of Scrapy. I tried twisted.names as alternative for high-load web spider. Generally it works. I didn't measured performance, but it seems there is no noticeable degradation at least. However, I faced with critical issue. 

Here is a stacktrace:
```python
/usr/local/lib/python2.7/dist-packages/twisted/names/dns.py(2680)_query()
-> writeMessage(m)
/usr/local/lib/python2.7/dist-packages/twisted/names/dns.py(2800)writeMessage()
-> self.writeMessage(m, address)
/usr/local/lib/python2.7/dist-packages/twisted/names/dns.py(2728)writeMessage()
-> self.transport.write(message.toStr(), address)
> /usr/local/lib/python2.7/dist-packages/twisted/internet/udp.py(286)write()
-> "write() only accepts IP addresses, not hostnames")
```
Dumping address I'm getting something like *('ns1.jiasule.net', 53)*. From [twisted documentation for UDP layer](https://twistedmatrix.com/documents/current/core/howto/udp.html) we have 
> Note that connect(), like write() will only accept IP addresses, not unresolved hostnames.

It seems Twisted UDP layer doesn't allow to twisted.names resolver to interact with DNS servers referenced by hostname. According to my experiment it mostly affects small networks DNS servers which dns response wasn't cached on higher levels.

Therefore, at the moment twisted.names.client.Resolver cannot be used for general purpose crawling.

**Comment by nyov** [comment]
What would you say about keeping the `CachingThreadedResolver` and `ThreadpoolAdjustingResolver` thing separate?
`DNSCACHE_SIZE` is still an improvement to the current code, whether the rest works or not.
I think renaming it to `ScrapyResolver` is less useful, that's less informational than the current name (about what it does).

**Comment by sibiryakov** [comment]
Why do we need so much ThreadedResolver customizations? I used ScrapyResolver, because it is general enough for handling all diversity of things we need. Is there any reason to separate threadpool adjustment logic to it's own class?

**Comment by nyov** [comment]
Well, from what kmike noted, it was my understanding that the threadpool adjustment in this PR doesn't work out, as is.
So the logical thing to me was taking the other changes about the DNSCACHE_SIZE and merging them by their own, as changes to the currently existing `CachingThreadedResolver` class.
If I made the wrong assumption there, I'm sorry and please ignore my comment.

**Comment by sibiryakov** [comment]
Guys, I have an idea. Let's implement thread pool adjustment somewhere else, outside of DNS resolver. E.g. setting can be REACTOR_THREADPOOL_MINSIZE. And pass the rest of changes to DNS resolver as they are now. What do you think? @kmike @nyov

**Comment by nyov** [comment]
I'm the wrong person to ask, my understanding there is limited - I don't even know if a reactor can have multiple threadpools assigned.
Though if we micromanage the threadpool we could in theory also say how many thread per service we want assigned (even if they aren't actually use-restricted to a specific scrapy service). So we might then define our threadpool should be the size of [DNS threads+Crawler threads+Pipeline threads+...] ? I hope that's not completely barking up the wrong tree here.

File: scrapy/resolver.py [hunk 2]
```diff
@@ -4,17 +4,28 @@
...
+        threadpool = self.reactor.getThreadPool()
+        threadpool.adjustPoolsize(maxthreads=settings.getint('DNS_MAX_THREADS'))
+        self.timeout = tuple(settings.getlist('DNS_TIMEOUT'))
+
+    def getHostByName(self, name, timeout=None):
+        if self.caching_enabled and name in dnscache:
             return defer.succeed(dnscache[name])
-        d = super(CachingThreadedResolver, self).getHostByName(name, timeout)
-        d.addCallback(self._cache_result, name)
+        if not timeout:
```
Comment (kmike) [inline scrapy/resolver.py]: Why isn't it `timeout is None`?

**Comment by kmike** [comment]
@nyov by default ThreadedResolver uses a common shared thread pool, and other Scrapy components also use shared threadpool. It is possible to create separate thread pools for each component; this could be a bit easier for Scrapy components and harder for builtin Twisted components.

**Comment by nyov** [comment]
Ah, that is good to know. And trying to micromanage threadpool would probably be digging too deep, and just having a global adjustment setting as @sibiryakov proposes is better, if we consider other technologies such as asyncio in the future and the 'other languages' support for spiders or even other scrapy components.

I don't know what this shared threadpool's size is and never had any issues with it, but if it's not currently adjustable that looks like a sensible option to have.

**Comment by sibiryakov** [comment]
Ok, guys I'll follow up with two new PRs: dns options foe existing threaded resolver and global option for thread pool size.
Considering GIL in python I think threading in python is makes sense only if threads are blocking on IO. If there're doing something computationally intensive like parsing, there will be no benefit.

**Comment by sibiryakov** [comment]
Closed in favor of https://github.com/scrapy/scrapy/pull/1120 and https://github.com/scrapy/scrapy/pull/1123

**Pull request closed without merging**